onnx_models/conftest.py

```diff
@@ -0,0 +1,355 @@
+    )
+    ret = subprocess.run(compile_cmd, shell=True, capture_output=True, cwd=cwd)
+    if ret.returncode != 0:
+        logger.error(f"Compilation of '{iree_module_path}' failed")
+        logger.error("iree-compile stdout:")
+        logger.error(ret.stdout.decode("utf-8"))
+        logger.error("iree-compile stderr:")
+        logger.error(ret.stderr.decode("utf-8"))
+        raise IreeCompileException(f"  '{iree_module_path.name}' compile failed")
+    return iree_module_path
+
+
+def run_iree_module(iree_module_path: Path, run_flags: List[str]):
+    cwd = THIS_DIR
+    run_args = ["iree-run-module", f"--module={iree_module_path.relative_to(cwd)}"]
+    run_args.extend(run_flags)
+    run_cmd = subprocess.list2cmdline(run_args)
+    logger.info(
+        f"Launching run command:\n"  #
+        f"  cd {cwd} && {run_cmd}"
+    )
+    ret = subprocess.run(run_cmd, shell=True, capture_output=True, cwd=cwd)
+    if ret.returncode != 0:
+        logger.error(f"Run of '{iree_module_path}' failed")
+        logger.error("iree-run-module stdout:")
+        logger.error(ret.stdout.decode("utf-8"))
+        logger.error("iree-run-module stderr:")
+        logger.error(ret.stderr.decode("utf-8"))
+        raise IreeRunException(f"  '{iree_module_path.name}' run failed")
+
+
+@pytest.fixture
+def compare_between_iree_and_onnxruntime():
+    def fn(
+        model_name: str,
+        model_url: str,
+    ):
+        if not ARTIFACTS_DIR.is_dir():
+            ARTIFACTS_DIR.mkdir(parents=True)
+        # TODO(scotttodd): group model artifacts into subfolders
+
+        # TODO(scotttodd): move to fixture with cache / download on demand
+        # TODO(scotttodd): extract name from URL?
+        # TODO(scotttodd): overwrite if already existing? check SHA?
+        original_onnx_path = ARTIFACTS_DIR / f"{model_name}.onnx"
+        if not original_onnx_path.exists():
+            urllib.request.urlretrieve(model_url, original_onnx_path)
+
+        # TODO(scotttodd): cache ONNX metadata and runtime results
+        upgraded_onnx_path = upgrade_onnx_model_version(original_onnx_path)
+
+        onnx_model_metadata = get_onnx_model_metadata(upgraded_onnx_path)
+        logger.debug("ONNX model metadata:")
+        logger.debug(onnx_model_metadata)
+
+        # Run through ONNX Runtime.
+        onnx_session = InferenceSession(upgraded_onnx_path)
+        output_names = [output["name"] for output in onnx_model_metadata["outputs"]]
+        inputs = {}
+        for input in onnx_model_metadata["inputs"]:
+            inputs[input["name"]] = input["input_data"]
+        onnx_results = onnx_session.run(output_names, inputs)
+
+        # Prepare inputs and expected outputs for running through IREE.
+        run_module_args = []
+        for input in onnx_model_metadata["inputs"]:
+            input_type = input["iree_type"]
+            input_data_path = input["input_data_path"]
+            run_module_args.append(f"--input={input_type}=@{input_data_path}")
```

Comment:
We have control over the inputs. As long as there aren't restrictions (like must be a multiple of 4, or two dynamic inputs must be equal), setting to 1 seems safe enough to start. We could also override per test case or add a flag that lets you override for the entire test suite.

I at least made it so the output shape comes from the inference session, since that value is out of out control.

Reply:
Yeah, I know many llms have dim params like `past_seq_len` and then elsewhere `past_seq_len + 1`, so there are definitely restrictions for some models. 

After looking more carefully at the setup (which is really nice, btw), I think it would be pretty feasible to just add an optional `dim_param_dict` as an input for `compare_between_iree_and_onnxruntime_fn`.

Reply:
In short, I think that could be added later as necessary.

Reply:
Yep. I just started with vision classification models but as more are added this test code will grow more features. Are there particular ONNX models that you know need that treatment? I can look at importing those next.